src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/script/KNNScoringSpace.java

```diff
@@ -0,0 +1,209 @@
+     * docs in an index is compatible.
+     *
+     * @param query Raw query object passed in to be validated and processed for the given similarity space
+     */
+    protected abstract void prepareQuery(Object query);
+
+    protected boolean isLongFieldType(MappedFieldType fieldType) {
+        return fieldType instanceof NumberFieldMapper.NumberFieldType
+                && ((NumberFieldMapper.NumberFieldType) fieldType).numericType() == LONG.numericType();
+    }
+
+    protected boolean isBinaryFieldType(MappedFieldType fieldType) {
+        return fieldType instanceof BinaryFieldMapper.BinaryFieldType;
+    }
+
+    protected boolean isKNNVectorFieldType(MappedFieldType fieldType) {
+        return fieldType instanceof KNNVectorFieldMapper.KNNVectorFieldType;
+    }
+
+    protected Long parseLongQuery(Object query) {
+        /*
+         * Because there is no way to specify the type of integral that is passed in during query, it is necessary to
+         * cast it to a Long here.
+         */
+        Long processedQueryLong;
+        if (query instanceof Integer) {
+            processedQueryLong = Long.valueOf((Integer) query);
+        } else if (query instanceof Long) {
+            processedQueryLong = (Long) query;
+        } else {
+            throw new IllegalArgumentException("Incompatible query_value for hamming space. query_value must " +
+                    "be either a Long or an Integer.");
+        }
+
+        return processedQueryLong;
+    }
+
+    protected BitSet parseBinaryQuery(Object query) {
+        return BitSet.valueOf(Base64.getDecoder().decode((String) query));
+    }
+
+    protected float[] parseKNNVectorQuery(Object query) {
+        float[] parsedQuery = KNNScoringUtil.convertVectorToPrimitive(query);
+        if (((KNNVectorFieldMapper.KNNVectorFieldType) fieldType).getDimension() != parsedQuery.length) {
+            KNNCounter.SCRIPT_QUERY_ERRORS.increment();
+            throw new IllegalStateException("[KNN] query vector and field vector dimensions mismatch. " +
+                    "query vector: " + parsedQuery.length + ", stored vector: " +
+                    ((KNNVectorFieldMapper.KNNVectorFieldType) fieldType).getDimension());
+        }
+        return parsedQuery;
+    }
+
+    /**
+     * Return the correct scoring script for a given query. The scoring script
+     *
+     * @param params Map of parameters
+     * @param field Fieldname
+     * @param lookup SearchLookup
+     * @param ctx ctx LeafReaderContext to be used for scoring documents
+     * @return ScoreScript for this query
+     * @throws IOException throws IOException if ScoreScript cannot be constructed
+     */
+    public abstract ScoreScript getScoreScript(Map<String, Object> params, String field, SearchLookup lookup,
+                                               LeafReaderContext ctx) throws IOException;
+
+    public static class L2 extends KNNScoringSpace {
```

Comment:
I believe L2 is Euclidean distance, may be name class as Euclidean?

Reply:
got it. i was concerned about having a class name with two characters. If it makes sense, i totally agree.